wait/awaitility.go

```diff
@@ -258,3 +268,149 @@ func containsClusterCondition(conditions []v1alpha1.ToolchainClusterCondition, c
+	// now, create the route for the service (if needed)
+	route := routev1.Route{}
+	if err := a.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+	}, &route); err != nil {
+		require.True(a.T, errors.IsNotFound(err), "failed to get route to access the '%s' service", service.Name)
+		route = routev1.Route{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: service.Namespace,
+				Name:      service.Name,
+			},
+			Spec: routev1.RouteSpec{
+				Port: &routev1.RoutePort{
+					TargetPort: intstr.FromString("http-metrics"),
+				},
+				TLS: &routev1.TLSConfig{
+					Termination: routev1.TLSTerminationEdge,
+				},
+				To: routev1.RouteTargetReference{
+					Kind: service.Kind,
+					Name: service.Name,
+				},
+			},
+		}
+		if err = a.Client.Create(context.TODO(), &route, &test.CleanupOptions{}); err != nil {
+			return route, err
+		}
+	}
+	return a.WaitForRouteToBeAvailable(route.Namespace, route.Name, endpoint)
+}
+
+// WaitForRouteToBeAvailable wais until the given route is available, ie, it has an Ingress with a host configured
+// and the endpoint is reachable (with a `200 OK` status response)
+func (a *SingleAwaitility) WaitForRouteToBeAvailable(ns, name, endpoint string) (routev1.Route, error) {
+	route := routev1.Route{}
+	// retrieve the route for the registration service
+	err := wait.Poll(a.RetryInterval, a.Timeout, func() (done bool, err error) {
+		if err = a.Client.Get(context.TODO(),
+			types.NamespacedName{
+				Namespace: ns,
+				Name:      name,
+			}, &route); err != nil {
+			if errors.IsNotFound(err) {
+				a.T.Logf("Waiting for creation of route '%s' in namespace '%s'...\n", name, ns)
+				return false, nil
+			}
+			return false, err
+		}
+		// assume there's a single Ingress and that its host will not be empty when the route is ready
+		if len(route.Status.Ingress) == 0 || route.Status.Ingress[0].Host == "" {
+			a.T.Logf("Waiting for availability of route '%s' in namespace '%s'...\n", name, ns)
+			return false, nil
+		}
+		// verify that the endpoint gives a `200 OK` response on a GET request
+		var location string
+		client := http.Client{
+			Timeout: time.Duration(1 * time.Second),
+		}
+		if route.Spec.TLS != nil {
+			location = "https://" + route.Status.Ingress[0].Host + endpoint
+			client.Transport = &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			}
+		} else {
+			location = "http://" + route.Status.Ingress[0].Host + endpoint
+		}
+		resp, err := client.Get(location)
+		if err, ok := err.(*url.Error); ok && err.Timeout() {
+			// keep waiting if there was a timeout: the endpoint is not available yet (pod is still re-starting)
+			a.T.Logf("Waiting for availability of route '%s' in namespace '%s' (endpoint timeout)...\n", name, ns)
+			return false, nil
+		} else if err != nil {
+			return false, err
+		}
+		defer func() {
+			_ = resp.Body.Close()
```

Comment:
I think it's not critical here but at least in previous golang version resp.Body.Close() without reading the body first did not actually release the file descriptor.

Reply:
oh, I thought this trick was what we needed to avoid keeping the file descriptor 🤦‍♂️ 
I'll check that

Reply:
refactored in 9d3e037: response body is read and closed (with `defer`) before checking the status code